compiler/resolution/lowerForalls.cpp

```diff
@@ -0,0 +1,898 @@
+            continue;
+          }
+          if (CallExpr* srcCall = toCallExpr(defSrc)) {
+            // Found it. Place it where our ForallStmt will go.
+            defMove->replace(srcCall->remove());
+            INT_ASSERT(currSym->firstSymExpr() == NULL); // no other refs to it
+            return srcCall;
+          }
+        }
+
+    // The AST is beyond our expectations. Bail out.
+    USR_FATAL(currSym, "only simple control flow is allowed in a procedure that returns an iterator for use in a forall loop");
+    break;
+  }
+
+  return NULL; //dummy
+}
+
+//
+// If the iterable expression calls something that is NOT
+// a parallel iterator proper, pre-process it and update
+// 'iterCall' and 'iterFn' to be the new iterable expression.
+//
+// Tests:
+//   library/packages/Collection/CollectionCounter.chpl
+//   library/standard/Random/deitz/test1D2D.chpl
+//   reductions/deitz/test_maxloc_reduce_wmikanik_bug2.chpl
+//
+static void handleIteratorForwarders(ForallStmt* fs,
+                                     CallExpr*& iterCall, FnSymbol*& iterFn)
+{
+  // These should have been replaced away in convertIteratorForLoopexpr().
+  INT_ASSERT(strncmp(iterFn->name, "_iterator_for_loopexpr", 22));
+
+  // Inline the forwarder, i.e. 'iterFn', like so:
+  //
+  //   forall x in iter1() do BODY;
+  //   proc iter1() { doSomething; return iter2(); }
+  //
+  //  ==>
+  //
+  //   doSomething;
+  //   forall x in iter2() do BODY;
+  //
+  // Correspondingly, update:
+  //   iterFn   <- iter2
+  //   iterCall <- the call iter2() in the forall
+
+  BlockStmt* fBody = copyFnBodyForInlining(iterCall, iterFn, fs);
+  fs->replace(fBody);
+
+  CallExpr* forwardee = stripReturnScaffolding(fBody);
+
+  forwardee->replace(fs);
+  iterCall->replace(forwardee);
+
+  iterCall = forwardee;
+  iterFn   = iterCall->resolvedFunction();
+
+  // Todo: handle the case where the new 'iterFn' is yet
+  // another forwarder. If so, repeat the same steps
+  // until we reach an iterFn that is a parallel iterator.
+}
+
+
+/////////// main driver ///////////
+
+// Remove supporting references in ShadowVarSymbols.
+// Otherwise flattenNestedFunction() will try to propagate them.
+static void clearUpRefsInShadowVars() {
+  forv_Vec(ShadowVarSymbol, svar, gShadowVarSymbols)
+    if (svar->inTree())
+      INT_ASSERT(false); // was: svar->removeSupportingReferences(); // wass - remove removeSupportingReferences()
+}
+
+/*
+We have created some nested task functions in expandTaskFn().
+Flatten them because the compiler will crash otherwise.
+
+Implementation considerations:
+* For some task functions, we create them explicitly in expandTaskFn().
+* Some others are created implicitly when we clone iterators that
+  contain such task functions.
+* The way flattenNestedFunctions() works as of this writing,
+  it is faster to flatten all task fns at once, rather than one by one.
+*/
+// Or we could do it for all iterators in or near removeUncalledIterators().
```

Comment:
I'm having some trouble following how the above comment relates to this function. It's removing task functions, not flattening them. Presumably they are already handled during forall inlining? Maybe the comment just needs updating?